lib/logger/lib/logger.ex

```diff
@@ -711,29 +714,43 @@ defmodule Logger do
   end
 
   @doc false
-  def __do_log__({level, mode, config, pdict}, chardata_or_fun, metadata)
-      when is_list(metadata) do
-    %{utc_log: utc_log?, truncate: truncate} = config
-    metadata = [pid: self()] ++ into_metadata(metadata, pdict)
-
-    case normalize_message(chardata_or_fun, metadata) do
-      {message, metadata} ->
-        tuple = {Logger, truncate(message, truncate), Logger.Utils.timestamp(utc_log?), metadata}
-
-        try do
-          notify(mode, {level, Process.group_leader(), tuple})
-          :ok
-        rescue
-          ArgumentError -> {:error, :noproc}
-        catch
-          :exit, reason -> {:error, reason}
-        end
+  def __do_log__({level, _mode, _config, _pdict}, fun, metadata)
+      when is_list(metadata) and is_function(fun, 0) do
+    level =
+      case level do
+        :warn -> :warning
+        level -> level
+      end
+    metadata = Map.new(metadata)
```

Comment:
Those conversions would be expensive. I would try to keep them as maps as much as possible. Only the ones passed at runtime as keyword lists would need to be converted to maps. I would even consider first changing the backends to accept to receive map metadata... I am now sure how much it would be considered a breaking change though.

Another idea is to pass the Logger metadata as a key inside the Erlang metadata... that would avoid key conflicts and what not and allow us to keep the preivous data type.

Reply:
This is only temporary, as I am still working on porting rest of the API to use maps within API and change map to keyword list only in compatibility layer just before passing it further to the backends.